src/wincng.c

```diff
@@ -1679,6 +1814,1204 @@ _libssh2_wincng_dsa_free(libssh2_dsa_ctx *dsa)
+ */
+
+int
+_libssh2_wincng_ecdh_create_key(IN LIBSSH2_SESSION *session,
+                                OUT _libssh2_wincng_ecdsa_key **privatekey,
+                                OUT unsigned char **encoded_publickey,
+                                OUT size_t *encoded_publickey_len,
+                                IN libssh2_curve_type curve)
+{
+    int result = LIBSSH2_ERROR_NONE;
+    NTSTATUS status;
+
+    BCRYPT_KEY_HANDLE key_handle = NULL;
+
+    /* Validate parameters */
+    if(curve >= _countof(_wincng_ecdsa_algorithms)) {
+        return LIBSSH2_ERROR_INVAL;
+    }
+
+    if(!_libssh2_wincng.hAlgECDH[curve]) {
+        return LIBSSH2_ERROR_INVAL;
+    }
+
+    if(!privatekey || !encoded_publickey || !encoded_publickey_len) {
+        return LIBSSH2_ERROR_INVAL;
+    }
+
+    *encoded_publickey = NULL;
+    *encoded_publickey_len = 0;
+
+    /* Create an ECDH key pair using the requested curve */
+    status = BCryptGenerateKeyPair(
+        _libssh2_wincng.hAlgECDH[curve],
+        &key_handle,
+        _wincng_ecdsa_algorithms[curve].key_length,
+        0);
+    if(!BCRYPT_SUCCESS(status)) {
+        result = _libssh2_error(
+            session,
+            LIBSSH2_ERROR_PUBLICKEY_PROTOCOL,
+            "Creating ECC key pair failed");
+        goto cleanup;
+    }
+
+    status = BCryptFinalizeKeyPair(key_handle, 0);
+    if(!BCRYPT_SUCCESS(status)) {
+        result = _libssh2_error(
+            session,
+            LIBSSH2_ERROR_PUBLICKEY_PROTOCOL,
+            "Creating ECDH key pair failed");
+        goto cleanup;
+    }
+
+    result = _libssh2_wincng_uncompressed_point_from_publickey(
+        session,
+        curve,
+        key_handle,
+        encoded_publickey,
+        encoded_publickey_len);
+    if(result != LIBSSH2_ERROR_NONE) {
+        result = _libssh2_error(
+            session,
+            LIBSSH2_ERROR_PUBLICKEY_PROTOCOL,
+            "Exporting ECDH key pair failed");
+    }
+
+    *privatekey = malloc(sizeof(_libssh2_wincng_ecdsa_key));
+    if(!*privatekey) {
+        result = LIBSSH2_ERROR_ALLOC;
+        goto cleanup;
+    }
+
+    (*privatekey)->curve = curve;
+    (*privatekey)->handle = key_handle;
+
+cleanup:
+    if(result != LIBSSH2_ERROR_NONE && key_handle) {
+        (void)BCryptDestroyKey(key_handle);
+    }
+
+    if(result != LIBSSH2_ERROR_NONE && *privatekey) {
+        free(*privatekey);
+    }
+
+    return result;
+}
+
+/*
+ * _libssh2_ecdsa_curve_name_with_octal_new
+ *
+ * Creates an ECDSA public key from an uncompressed point.
+ */
+
+int
+_libssh2_wincng_ecdsa_curve_name_with_octal_new(
+    OUT _libssh2_wincng_ecdsa_key **key,
+    IN const unsigned char *publickey_encoded,
+    IN size_t publickey_encoded_len,
+    IN libssh2_curve_type curve)
+{
+    int result = LIBSSH2_ERROR_NONE;
+
+    BCRYPT_KEY_HANDLE publickey_handle;
+    _libssh2_ecdsa_point publickey;
+
+    /* Validate parameters */
+    if(curve >= _countof(_wincng_ecdsa_algorithms)) {
```

Comment:
Can we replace `_countof()` with our macro `ARRAY_SIZE()` to stay within ANSI C89?

https://github.com/libssh2/libssh2/blob/3a6ab70dcfeb87a3acbb5af7ea6fc783c3981e04/src/libssh2_priv.h#L1239

Also for the other uses below.

Reply:
fixed